### PR TITLE
feat(sdk): implement otlp monitoring exporter

### PR DESCRIPTION
**Pull request by bojiang**
Now bentoml monitoring can work with any collection libs that support otlp

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -156,6 +156,11 @@ tracing-jaeger = ["opentelemetry-exporter-jaeger==1.14.0"]
 tracing-zipkin = ["opentelemetry-exporter-zipkin==1.14.0"]
 tracing-otlp = ["opentelemetry-exporter-otlp==1.14.0"]
 
+monitoring = [
+  "bentoml[monitoring-otlp]",
+]
```
Comment (aarnphm) [inline pyproject.toml:161]: Hmm, idk if we should add this. I prefer explicitly declare external dependencies for these.

 The reason we have a `bentoml[tracing]` is for backward compatibility.

File: src/bentoml/_internal/monitoring/api.py [hunk 2]
```diff
@@ -253,7 +253,12 @@ def monitor(
     Context manager for monitoring.
 
     :param name: name of the monitor
-    :param monitor_class: class of the monitor
+    :param monitor_class: class of the monitor, can be a string or a class
```
Comment (aarnphm) [inline src/bentoml/_internal/monitoring/api.py:256]: can you convert this to Google-style docstring?

File: src/bentoml/_internal/monitoring/otlp.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+import os
+import random
+import typing as t
+import logging
+import datetime
+import collections
+import logging.config
+from typing import TYPE_CHECKING
+
+from opentelemetry.sdk._logs import LoggerProvider
```
Comment (aarnphm) [inline src/bentoml/_internal/monitoring/otlp.py:12]: Why do we have to use a private library like this? Don't opentelemetry has a public SDK for logs?

Comment (aarnphm) [reply src/bentoml/_internal/monitoring/otlp.py:12]: Seems like they dont 😢

**Comment by aarnphm** [comment]
Another thing can you also add the additional dependency to `--extra-requires`, which is defined here https://github.com/bentoml/BentoML/blob/bd547bc5405a151aeed5091ce69f6af1a4e6ee4d/src/bentoml/bentos.py#L441?

**Comment by aarnphm** [comment]
Also remember to add `monitoring-otlp` to `all` as well

**Review by aarnphm — APPROVED** [review]

File: src/bentoml/_internal/monitoring/otlp.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+            resource=Resource.create(
+                {
+                    "service.name": f"{component_context.bento_name}:{self.name}",
+                    "service.instance.id": "{component_context.bento_version}",
+                }
+            ),
+        )
+        set_logger_provider(self.logger_provider)
+
+        if self.endpoint is not None:
```
Comment (aarnphm) [inline src/bentoml/_internal/monitoring/otlp.py:141]: We shouldn't override user env if they set themselves right?

Comment (bojiang) [reply src/bentoml/_internal/monitoring/otlp.py:141]: It has the same behavior as open telemetry implementation now.

Comment (judahrand) [reply src/bentoml/_internal/monitoring/otlp.py:141]: This isn't the behavior now - I don't know if it was at the time: https://github.com/open-telemetry/opentelemetry-python/blob/ea15f72f6678ff23c86511c095094cb3e4bafc44/exporter/opentelemetry-exporter-otlp-proto-http/src/opentelemetry/exporter/otlp/proto/http/_log_exporter/__init__.py#L75-L94

**Pull request merged**